server/app/middlewares/websocket_backend.rb

```diff
@@ -66,43 +87,84 @@ def call(env)
   # @param [Faye::WebSocket] ws
   # @param [Rack::Request] req
   def on_open(ws, req)
+    # check grid
     grid = Grid.find_by(token: req.env['HTTP_KONTENA_GRID_TOKEN'].to_s)
-    if !grid.nil?
-      node_id = req.env['HTTP_KONTENA_NODE_ID'].to_s
+    unless grid
+      logger.error 'invalid grid token, closing connection'
+      ws.close(4001)
+      return
+    end
 
-      logger.info "node #{node_id} opened connection"
+    # check host node
+    node_id = req.env['HTTP_KONTENA_NODE_ID'].to_s
+    labels = req.env['HTTP_KONTENA_NODE_LABELS'].to_s.split(',')
 
-      node = grid.host_nodes.find_by(node_id: node_id)
-      labels = req.env['HTTP_KONTENA_NODE_LABELS'].to_s.split(',')
-      unless node
-        node = grid.host_nodes.create!(node_id: node_id, labels: labels)
-      end
+    node = grid.host_nodes.find_by(node_id: node_id)
 
-      node_plugger = Agent::NodePlugger.new(grid, node)
-      client = {
-          ws: ws,
-          id: node_id.to_s,
-          node_id: node.id,
-          grid_id: grid.id,
-          created_at: Time.now
-      }
-      @clients << client
-
-      agent_version = req.env['HTTP_KONTENA_VERSION'].to_s
-      unless self.valid_agent_version?(agent_version)
-        logger.error "version mismatch: server (#{Server::VERSION}), node (#{agent_version})"
-        node_plugger.send_master_info
-        self.handle_invalid_agent_version(ws, node)
-        return
-      end
+    unless node
+      node = grid.host_nodes.create!(node_id: node_id, labels: labels)
+    end
 
-      EM.defer { node_plugger.plugin! }
-    else
-      logger.error 'invalid grid token, closing connection'
-      ws.close(4001)
+    # check version
+    agent_version = req.env['HTTP_KONTENA_VERSION'].to_s
+
+    unless self.valid_agent_version?(agent_version)
+      handle_invalid_agent_version(ws, node, agent_version)
+      return
     end
+
+    # check clock
+    connected_at = Time.parse(req.env['HTTP_KONTENA_CONNECTED_AT'])
+    connected_dt = Time.now - connected_at
+
+    if connected_dt > PING_TIMEOUT + CLOCK_SKEW || connected_dt < -CLOCK_SKEW
+      handle_invalid_agent_clock(ws, node, connected_at, connected_dt)
+      return
+    end
+
+    logger.info "node #{node} agent version #{agent_version} connected at #{connected_at}, #{'%.2fs' % connected_dt} ago"
+
+    client = {
+        ws: ws,
+        id: node_id.to_s,
+        node_id: node.id,
+        grid_id: grid.id,
+        created_at: Time.now,
+        connected_at: connected_at,
+    }
+
+    plugin_client(client, node)
   rescue => exc
-    logger.error "#{exc.class.name}: #{exc.message}"
+    logger.error exc
+  end
+
+  # @param [Faye::WebSocket] ws
+  # @param [HostNode] node
+  def handle_invalid_agent_version(ws, node, version)
+    logger.warn "node #{node} agent version #{version} is not compatible with server version #{Server::VERSION}"
+
+    ws.close(4010, "agent version #{version} is not compatible with server version #{Server::VERSION}")
```

Comment:
This is missing the `node_plugger.send_master_info` + `EventMachine::Timer.new(1) do ws.close` that was there before...